### PR TITLE
docs: Clarify session modes and give SDK suggestions

### PR DESCRIPTION
**Pull request by Swatinem**
Still an open question:

* how should `start_session` behave with regards to ending the previous session. In case you have application-mode or request-mode.

### CODE REVIEW

**Review by mitsuhiko — APPROVED** [review]

File: src/docs/sdk/sessions.mdx [hunk 1]
```diff
@@ -239,26 +239,66 @@ sessions over time.
...
+Conceptually speaking, the session is a concern of the `Hub`, and unlike scopes,
+sessions should not be nested. When any kind of event happens, there should be
+only one unambiguous session which keeps track of the error count.
+
+When considering the flow of events through the SDK, from the static
+`capture_event` function, through the thread local `Hub`, and into the
+`Client::capture_event(event, scope)` method; depending on the internal
+implementation details of the SDK, it might make sense to attach the session to
+the `Scope`, which would make it possible for the `Client` to bundle an event
+and a session update into a single envelope to be sent to the sentry backend.
```
Comment (rhcarvalho) [inline src/docs/sdk/sessions.mdx:280]:
```suggestion
and a session update into a single envelope to be sent to the Sentry backend.
```

File: src/docs/sdk/sessions.mdx [hunk 2]
```diff
@@ -239,26 +239,66 @@ sessions over time.
...
+
+### Session Updates and when to send Updates upstream
+
+For all SDKs the current session shall be automatically updated whenever data is
+captured at a similar place where `apply_to_scope` is called to increase the
+`error` count, or update the session based on the distinct ID / user ID.
+
+SDKs should generally aim to decrease the number of envelopes send upstream.
```
Comment (rhcarvalho) [inline src/docs/sdk/sessions.mdx:288]:
```suggestion
SDKs should generally aim to decrease the number of envelopes sent upstream.
```

File: src/docs/sdk/sessions.mdx [hunk 3]
```diff
@@ -239,26 +239,66 @@ sessions over time.
 Generally speaking there are two separate modes for health reporting that SDKs
 can use. One is very short lived sessions, the other are user attended sessions.
 
-**Short lived sessions** *(req-mode)*
+**Short lived sessions** *(server-mode / request-mode)*
 
-These are sessions that correspond to a single HTTP request or RPC call.
+These are sessions roughly correspond to a HTTP requests or RPC calls in a
```
Comment (lobsterkatie) [inline src/docs/sdk/sessions.mdx:244]:
```suggestion
These sessions roughly correspond to HTTP requests or RPC calls in a
```

File: src/docs/sdk/sessions.mdx [hunk 4]
```diff
@@ -239,26 +239,66 @@ sessions over time.
...
+server setting.
 
-- high in volume
+- high in volume, typically one session for each request
+- the number of sessions is usually higher than the number of events
```
Comment (lobsterkatie) [inline src/docs/sdk/sessions.mdx:248]: I'm not 100% clear what you mean here. More sessions than update events? More sessions than error events?

Or, perhaps a better question, what's the big idea you're trying to get across?

File: src/docs/sdk/sessions.mdx [hunk 5]
```diff
@@ -239,26 +239,66 @@ sessions over time.
...
+These are sessions that are more corresponding to an actual user session or
+application run. This is what you would see in a web browser, mobile world, 
+command line application or similar.
 
-- session duration typically in the minutes
-- timing information is useful
-- typically send in two or more session updates (start and end)
+- typically just a single session from application start to quit
+- there are usually fewer sessions than events
+- sessions spanning multiple hubs / threads
```
Comment (lobsterkatie) [inline src/docs/sdk/sessions.mdx:260]:
```suggestion
- sessions span multiple hubs / threads
```

File: src/docs/sdk/sessions.mdx (hunk 4 above)
Comment (Swatinem) [reply src/docs/sdk/sessions.mdx:248]: I mean the `capture_{message,event,exception}` events. Is there another term that makes it more clear?

**Pull request merged**